icefoganalytics/travel-authorization
# Issue 102: Standardize Travel Purpose Table and Front-End Wiring

# Context

In preparation of travel authorization management interface, I need to clean up some code.
This PR vastly reduces the complexity of using the "travel purpose/purpose" model.

# Implementation

Standardize `travelPurpose` table to `travel_purposes` and add timestamp columns.
Use new standard for Vuex pattern.

<details><summary><code>web/src/store/travel-purposes.js</code></summary>
<p>

```js
import travelPurposesApi from "@/api/travel-purposes-api"

import { withGettersFromState } from "@/utils/vuex-utils"

const state = {
  items: [],
  isLoading: false,
  isErrored: false,
  isCached: false,
}

const getters = withGettersFromState(state, {})

const actions = {
  async ensure({ commit, state, dispatch }) {
    if (state.isCached) {
      return state.items
    }

    commit("SET_IS_CACHED", false)
    return dispatch("fetch")
  },
  async fetch({ commit }) {
    commit("SET_IS_LOADING", true)
    try {
      const { travelPurposes } = await travelPurposesApi.list()
      commit("SET_IS_ERRORED", false)
      commit("SET_ITEMS", travelPurposes)
      commit("SET_IS_CACHED", true)
      return travelPurposes
    } catch (error) {
      console.error("Failed to fetch travel purposes:", error)
      commit("SET_IS_ERRORED", true)
      throw error
    } finally {
      commit("SET_IS_LOADING", false)
    }
  },
}

const mutations = {
  SET_ITEMS(state, value) {
    state.items = value
  },
  SET_IS_LOADING(state, value) {
    state.isLoading = value
  },
  SET_IS_ERRORED(state, value) {
    state.isErrored = value
  },
  SET_IS_CACHED(state, value) {
    state.isCached = value
  },
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
}
```

</p>
</details> 

# Testing Instructions

1. Boot the app via `dev up`.
2. Log in to the app at http://localhost:8080.
3. Go to the "My Travel Requests" page via the top drop down nav.
4. Create a new travel authorization via the "+ Travel Authorization" button.
5. Check that you can add a travel purpose to the new travel form.
6. Fill in the form and submit it.
7. Check that the purpose displays correctly in the read-only view of the travel request.
